File: algorithm/python/BigSquare1915/bigsquare_1915_p.py
import sys
input = sys.stdin.readline

N, M = map(int, input().split())
map = [[0] * M for _ in range(N)]

for i in range(N):
    inputs = list(input().strip())
    for j in range(M):
        map[i][j] = int(inputs[j])

max_cnt = 0

for i in range(N):
    for j in range(M):
        if (map[i][j] != 0):
            
            map_im_jm = map_im_j = map_i_jm = 0    

            map[i][j] = min(map[i-1][j-1], min(map[i-1][j], map[i][j-1])) + 1
            max_cnt = max(max_cnt, map[i][j])

print(max_cnt*max_cnt)

File: algorithm/java/Java_Algorithm/src/BigSquare1915/BigSquare1915p.java
package BigSquare1915;

import java.io.*;
import java.util.*;

public class BigSquare1915p {
    static int N,M;
    static int[][] map;
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer stk = new StringTokenizer(br.readLine(), " ");

        N = Integer.parseInt(stk.nextToken());
        M = Integer.parseInt(stk.nextToken());
        
        map = new int[N][M];

        for (int i = 0; i < N; i++) {
            String[] inputs = br.readLine().split("");

            for (int j = 0; j < M; j++) {
                map[i][j] = Integer.parseInt(inputs[j]);
            }
        }   

        int maxCnt = 0;

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (map[i][j] != 0) {
                    
                    int im = i - 1;
                    int jm = j - 1;

                    int mapImJm = (im < 0 || jm < 0) ? 0 : map[i-1][j-1];
                    int mapImJ = (im < 0) ? 0 : map[i-1][j];
                    int mapIJm = (jm < 0) ? 0 : map[i][j-1];
                    
                    map[i][j] = Math.min(mapImJm, Math.min(mapImJ, mapIJm)) + 1;
                    maxCnt = Math.max(maxCnt, map[i][j]);
                }
            }
        }

        bw.write(maxCnt*maxCnt + "");
        bw.close();
        br.close();
    }
}
